proposed/logging-generator.md

```diff
@@ -0,0 +1,496 @@
+    public partial void LogWithCustomEventName();
+
+    [LoggerMessage(10, "Welcome to {city} {province}!")]
+    public partial void LogWithDynamicLogLevel(string city, LogLevel level, string province);
+
+    [LoggerMessage(2, LogLevel.Trace)]
+    public partial void LogWithNoTemplate(string key1, string key2);
+
+    public void TestLogging()
+    {
+        LogEmptyMessage();
+        LogWithCustomEventName();
+        LogWithDynamicLogLevel("Vancouver", LogLevel.Warning, "BC");
+        LogWithDynamicLogLevel("Vancouver", LogLevel.Information, "BC");
+        LogWithNoTemplate("value2", "value2");
+    }
+}
+```
+output to `TestLogging()` using SimpleConsole:
+```
+trce: LoggingExample[1]
+      {}
+trce: LoggingExample[9]
+      Fixed message
+warn: LoggingExample[10]
+      Welcome to Vancouver BC!
+info: LoggingExample[10]
+      Welcome to Vancouver BC!
+trce: LoggingExample[2]
+      {"key1":"value2","key2":"value2"}
+```
+
+### Diagnostics
+
+This [gist](https://gist.github.com/maryamariyan/a1ab553bedb26b9886fbc2740ee9e954) shows 20 diagnostic messages that the generator can produce alongside use cases for each
+
+Using SYSLIBXXXX format as the diagnostic IDs. The gist also shows diagnostic categories against each sample.
+
+### Q & A - The source generator:
+
+- Question: Why would we want to enforce event ID uniqueness checks if existing log APIs still allow for using default event ID values? For example with `LogInformation` we have two overloads, one taking `EventId` and the other one not. But the proposal with `LoggerMessageAttribute` only requires taking an `EventId`.
+
+Answer: The enforced restriction added via `LoggerMessage` attribute aims at providing best practices for library authors more so than it does for app developers, who in most cases do not care about event IDs.
+
+This is a great feedback. In order to make this approach less restrictive, it would make sense to also allow `LoggerMessageAttribute` skip taking event IDs in combination with an analyzer that generates an error to recommend best practices while allowing app developers to cut corners for productivity. When turned off, the event IDs would default to zero or implicit event IDs would get generated.
+
+## Supporting string interpolation builder in logging
+
+This section describes using a language feature (called string interpolation builder) to solve the problem of strongly-typing code paths for ILogger messages.
+
+### Usage examples / API Proposal:
+
+As opposed to writing:
+
+```csharp
+[LoggerMessage(0, LogLevel.Information, "Hello `{name}`")]
+public static partial void LogName(this ILogger logger, string name, Exception exception);
+```
+
+described in the previous section, the usage would look like:
+
+```csharp
+public static void LogName(this ILogger logger, string name, Exception exception) =>
+    logger.Log(LogLevel.Information, new EventId(0), $"Hello `{name}`", exception, (s, e) => s.ToString());
```

Comment:
Not sure this works as is without adding new overloads to Logging to support the builder specifically. Essentially imagine we have a method like this:

```C#
void Log<T>(T state) => Console.WriteLine(state.ToString());
```
It has no idea what T is supposed to be when I do this:

```C#
string name = "David";
Log($"Hello {name}"); // This will call string.Format
```

I assume we'd need an overload like this:

```C#
void Log(CustomLoggerParamsBuilder loggingBuilder) => Console.WriteLine(loggingBuilder.ToString());
```

Now the compiler can perform target typing.

@stephentoub and @333fred

Reply:
> `Log($"Hello {name}"); // This will call string.Format`

Well, that would presumably call the `InterpolatedStringBuilder` in the framework, but the end result is the same, it's not your custom builder type. If you want to use your own builder, you need to have an overload that takes it instead of a string.

Reply:
Yep. The point was to call out that we'd need a new overload.

Reply:
Note that if you have both, the builder-type overload should be preferred. So it's not a problem if you have an overload that does take a string.

Reply:
It's possible to prevent fallback to string with the appropriate methods on the builder as well right?

Reply:
> It's possible to prevent fallback to string with the appropriate methods on the builder as well right?

The way to do that would be to not have an overload that takes a string.